dynawo/sources/Models/CPP/Controls/Shunts/ModelCentralizedShuntsSectionControl/DYNModelCentralizedShuntsSectionControl.cpp

```diff
@@ -0,0 +1,369 @@
+        if (doubleNotEquals(UMonitoredPu, minValue) &&
+          UMonitoredPu < minValue &&
+          sections0_[s] < SectionsMax_[s]) {
+          gLocal_[0] = ROOT_UP;
+          gLocal_[1] = ROOT_DOWN;
+          gLocal_[2] = ((t - whenUp_) > tNext_ || doubleEquals((t - whenUp_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+          changingShunt = s;
+          break;
+        } else if (doubleNotEquals(UMonitoredPu, maxValue) &&
+          UMonitoredPu > maxValue &&
+          sections0_[s] > SectionsMin_[s]) {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_UP;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ((t - whenDown_) > tNext_ || doubleEquals((t - whenDown_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          changingShunt = s;
+          break;
+        } else {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_DOWN;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+        }
+      }
+    } else {
+      for (int s = 0; s < nbShunts_; ++s) {
+        minValue = zLocal_[s] - DeadbandsUPu_[s];
+        maxValue = zLocal_[s] + DeadbandsUPu_[s];
+        if (doubleNotEquals(UMonitoredPu, minValue) &&
+          UMonitoredPu < minValue &&
+          sections0_[s] > SectionsMin_[s]) {
+          gLocal_[0] = ROOT_UP;
+          gLocal_[1] = ROOT_DOWN;
+          changingShunt = s;
+          gLocal_[2] = ((t - whenDown_) > tNext_ || doubleEquals((t - whenDown_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+        } else if (doubleNotEquals(UMonitoredPu, maxValue) &&
+          UMonitoredPu > maxValue &&
+          sections0_[s] < SectionsMax_[s]) {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_UP;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ((t - whenUp_) > tNext_ || doubleEquals((t - whenUp_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          changingShunt = s;
+        } else {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_DOWN;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+        }
+      }
+    }
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalZ(const double t) {
+    if (!isSelf_) {
+      if (gLocal_[0] == ROOT_UP) {
+        whenUp_ = t;
+        whenDown_ = VALDEF;
+      }
+      if (gLocal_[1] == ROOT_UP) {
+        whenDown_ = t;
+        whenUp_ = VALDEF;
+      }
+      if (gLocal_[2] == ROOT_UP) {
+        if (doubleNotEquals(lastTime_, t)) {
+          zLocal_[changingShunt + nbShunts_] = sections0_[changingShunt] + 1;
+          sections0_[changingShunt] += 1;
+          changingShunt = -1;
+          lastTime_ = t;
+        }
+      }
+      if (gLocal_[3] == ROOT_UP) {
+        if (doubleNotEquals(lastTime_, t)) {
+          zLocal_[changingShunt + nbShunts_] = sections0_[changingShunt] - 1;
+          sections0_[changingShunt] -= 1;
+          changingShunt = -1;
+          lastTime_ = t;
+        }
+      }
```

Comment:
else if cannot be everywhere because meaning of gLocal_[0] and gLocal[1] is different from gLocal_[2] and gLocal_[3].
Else if can olnly be added between gLocal_[0]/ gLocal_[1] and gLocal_[2]/gLocal_[3]